### PR TITLE
Redesign SPI factory to make them into one class (#1537)

### LINKED ISSUE #1537: Redesign SPI factory to make them into one class
The current factory using SPI are separate into multiple classes.
It is better redesign and abstract them into one class.

The SPI factory class are:

- JobItemExecutorFactory
- JobShardingStrategyFactory
- JobExecutorServiceHandlerFactory
- JobErrorHandlerFactory
- ElasticJobListenerFactory

### PR DESCRIPTION
**Pull request by wwj-go**
Fixes #1537.

Changes proposed in this pull request:

Redesign SPI factory to make logic about their TypedSPI into ElasticJobServiceLoader. The SPI factory class are:
- `JobShardingStrategyFactory`
- `JobExecutorServiceHandlerFactory`
- `JobErrorHandlerFactory`
- `ElasticJobListenerFactory`

The situation of `JobItemExecutorFactory `is a bit complicated and will be completed in subsequent work.

### CODE REVIEW

**Comment by wwj-go** [comment]
@TeslaCN , I’m sorry that I closed the PR just now due to a mistake, and I resubmitted another.

File: elasticjob-infra/elasticjob-infra-common/src/main/java/org/apache/shardingsphere/elasticjob/infra/spi/ElasticJobServiceLoader.java [hunk 1]
```diff
@@ -76,4 +81,58 @@ private static Object newServiceInstance(final Class<?> clazz) {
...
+     * Get cached instance.
+     *
+     * @param typedService service type
+     * @param type         specific service type
+     * @param <T>          specific type of service
+     * @return cached service instance
+     */
+    public static <T extends TypedSPI> T getCachedInstance(final Class<T> typedService, final String type) {
+        T instance = TYPED_SERVICES.containsKey(typedService) ? (T) TYPED_SERVICES.get(typedService).get(type) : null;
+        if (instance == null) {
```
Comment (TeslaCN) [inline elasticjob-infra/elasticjob-infra-common/src/main/java/org/apache/shardingsphere/elasticjob/infra/spi/ElasticJobServiceLoader.java]: Have constants on the left and variable on the right in = and equals conditional expressions; Have variable on the left and constants on the right in greater than and less than conditional expressions.
Ref https://shardingsphere.apache.org/community/en/contribute/code-conduct/

File: elasticjob-infra/elasticjob-infra-common/src/main/java/org/apache/shardingsphere/elasticjob/infra/spi/ElasticJobServiceLoader.java [hunk 2]
```diff
@@ -76,4 +81,58 @@ private static Object newServiceInstance(final Class<?> clazz) {
...
+     * New typed instance.
+     *
+     * @param typedService service type
+     * @param type         specific service type
+     * @param <T>          specific type of service
+     * @return specific typed service instance
+     */
+    public static <T extends TypedSPI> T newTypedServiceInstance(final Class<T> typedService, final String type) {
+        Class<?> instanceClass = TYPED_SERVICE_CLASSES.containsKey(typedService) ? TYPED_SERVICE_CLASSES.get(typedService).get(type) : null;
+        if (instanceClass == null) {
```
Comment (TeslaCN) [inline elasticjob-infra/elasticjob-infra-common/src/main/java/org/apache/shardingsphere/elasticjob/infra/spi/ElasticJobServiceLoader.java]: Have constants on the left and variable on the right in = and equals conditional expressions; Have variable on the left and constants on the right in greater than and less than conditional expressions.
Ref https://shardingsphere.apache.org/community/en/contribute/code-conduct/

**Review by TeslaCN — CHANGES_REQUESTED** [review]

**Comment by TeslaCN** [comment]
@wwj-go Thanks for you contribution. You can push commits instead of opening another PR.

**Comment by wwj-go** [comment]
@TeslaCN Thanks for your suggestions. 
As a novice in the community, your positive responses and sincere suggestions made me feel very warm and made me full of confidence in the community.
 I have push a commit  fixed the problem you mentioned above.

> @wwj-go Thanks for you contribution. You can push commits instead of opening another PR.

**Comment by jiang2015** [comment]
It is good to learn how to use java 0OPTIONAL.

**Comment by wwj-go** [comment]
About OPTIONAL, can you give some more specific suggestions in conjunction with this PR?
> It is good to learn how to use java 0OPTIONAL.

**Review by terrymanu — APPROVED** [review]

**Pull request merged**